### PR TITLE
Enhance docs

### PR DESCRIPTION
**Pull request by dpordomingo**
caused by https://github.com/src-d/lookout/issues/330
blocks https://github.com/src-d/lookout/pull/398

I did:
- added the `.github/CODE_OF_CONDUCT.md` file,
- moved the example analyzers into `examples` dir,
- fixed the `LICENSE.md` file, and created the `NOTICE.md` file,
- added the `DCO` file,
- added the `CONTRIBUTING.md` file
- added the `make toc` rule into the `Makefile,
- enhanced the `README.md` considering that **source{d} lookout** docs will be in `src-d/lookout` repo,

### CODE REVIEW

File: NOTICE.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+Copyright (C) 2018 Sourced Technologies, S.L.
```
Comment (smacker) [inline NOTICE.md]: I don't see this file in our guide. Could you please point why do we need it?

File: README.md [hunk 2]
```diff
@@ -1,117 +1,121 @@
...
+- pre-generated libraries code for [Golang](./pb) and [Python](./python):
+  - an easy **access to the DataService API though a gRPC** service to lookout analyzers, so they won't need to deal with actual Git repositories, UAST extraction, programming language detection, etc,
+  - low-level **helpers to workaround some protobuf/gRPC caveats**,
+- quickstart [examples](./examples) of an Analyzer that detects language and number of functions (written in Go and in Python).
+
+
+# Caveats
+
+When the gRPC client and servers are being used, it is needed:
+- to set [max gRPC message size](https://github.com/grpc/grpc/issues/7927); using lookout-sdk helpers it can be done:
```
Comment (smacker) [inline README.md]: maybe we need to explain why we change defaults. (UAST)

File: NOTICE.md (hunk 1 above)
Comment (dpordomingo) [reply NOTICE.md]: Not yet merged, but agreed after https://github.com/src-d/guide/pull/305#issuecomment-436596521
It's also added as part of the `project template` by https://github.com/src-d/guide/pull/273 (not yet merged, but almost there)

File: README.md [hunk 3]
```diff
@@ -1,117 +1,121 @@
-lookout SDK [![GoDoc](https://godoc.org/gopkg.in/src-d/lookout-sdk.v0?status.svg)](https://godoc.org/gopkg.in/src-d/lookout-sdk.v0/pb) [![PyPI version](https://badge.fury.io/py/lookout-sdk.svg)](https://pypi.org/project/lookout-sdk/) [![Build Status](https://travis-ci.org/src-d/lookout-sdk.svg)](https://travis-ci.org/src-d/lookout-sdk)
------------
+<h1 align="center">
```
Comment (smacker) [inline README.md]: according to our CTO html in README should be avoided:
https://github.com/src-d/backlog/issues/1318#issuecomment-435921226

File: README.md [hunk 4]
```diff
@@ -1,117 +1,121 @@
...
+      * [Community](#community)
+   * [Code of Conduct](#code-of-conduct)
+   * [License](#license)
+
+<!-- Added by: david, at: 2018-12-07T15:43+01:00 -->
+
+<!--te-->
+
+
+# What does the SDK provide?
```
Comment (carlosms) [inline README.md]:
```suggestion
# What Does the SDK Provide?
```

File: README.md [hunk 5]
```diff
@@ -1,117 +1,121 @@
...
+
+_For the complete documentation of **source{d} Lookout**, please take a look into [http://docs.sourced.tech/lookout](http://docs.sourced.tech/lookout)._
```
Comment (carlosms) [inline README.md]:
```suggestion
_For the complete documentation of **source{d} Lookout**, please take a look at [http://docs.sourced.tech/lookout](http://docs.sourced.tech/lookout)._
```

File: README.md [hunk 6]
```diff
@@ -1,117 +1,121 @@
...
+
+_For detailed information about the different parts of Lookout, and how they interact you can go to the [lookout architecture guide](https://github.com/src-d/lookout/blob/master/docs/architecture.md)._
+
+**lookout-sdk** provides:
+
+- **proto [definitions](./proto)**,
```
Comment (carlosms) [inline README.md]: I'd remove the trailing `,`, it looks strange to me

File: README.md [hunk 7]
```diff
@@ -1,117 +1,121 @@
...
+- pre-generated libraries code for [Golang](./pb) and [Python](./python):
```
Comment (carlosms) [inline README.md]:
```suggestion
- pre-generated libraries for [Golang](./pb) and [Python](./python), offering:
```

File: README.md [hunk 8]
```diff
@@ -1,117 +1,121 @@
...
+  - an easy **access to the DataService API though a gRPC** service to lookout analyzers, so they won't need to deal with actual Git repositories, UAST extraction, programming language detection, etc,
```
Comment (carlosms) [inline README.md]: This sounds a bit like the lookout analyzers will be serving that DataService
```suggestion
  - an easy **access to the DataService API though a gRPC service**. Lookout will take care of dealing with Git repositories, UAST extraction, programming language detection, etc
```

File: README.md [hunk 9]
```diff
@@ -1,117 +1,121 @@
...
+  - low-level **helpers to workaround some protobuf/gRPC caveats**,
```
Comment (carlosms) [inline README.md]:
```suggestion
  - low-level **helpers to work around some protobuf/gRPC caveats**,
```

File: README.md [hunk 10]
```diff
@@ -1,117 +1,121 @@
...
+- quickstart [examples](./examples) of an Analyzer that detects language and number of functions (written in Go and in Python).
+
+
+# Caveats
+
+When the gRPC client and servers are being used, it is needed:
```
Comment (carlosms) [inline README.md]: This sounds like you have an option to use or not gRPC. What about
```
For the gRPC client and server please to follow these requirements:
- set...
- support...
```

File: README.md (hunk 6 above)
Comment (smacker) [reply README.md]: I also noticed it but I thought it's just me

File: README.md (hunk 3 above)
Comment (dpordomingo) [reply README.md]: I also read https://github.com/src-d/guide/pull/292#issuecomment-436597861 
>_Engine has this version because is a product [...]_

So since **source{d} lookout** is also a product, I just coppied the header from **source{d} engine**.

Has @src-d/product any guidelines about product's `README.md`?

File: README.md (hunk 10 above)
Comment (dpordomingo) [reply README.md]:
```suggestion
For the gRPC client and server please to follow these requirements:
```

File: README.md (hunk 2 above)
Comment (dpordomingo) [reply README.md]: I agree.

Just to confirm that I understood it, I'd reword as:
>_to set a shared maximum gRPC message size in gRPC servers and clients (to avoid hitting different gRPC limits [see [grpc/grpc#7927](https://github.com/grpc/grpc/issues/7927)]); the `Analyzer` service should do so when dialing `DataService` and serving `NotifyReviewEvent` and `NotifyPushEvent`; using lookout-sdk helpers it can be done:

**Comment by dpordomingo** [comment]
Thanks for your reviews; I addressed all your suggestions and posted two questions.
(DCO problem now because accepting suggestions; I'll fix it once the PR has been accepted)

File: README.md (hunk 3 above)
Comment (smacker) [reply README.md]: I think we need input from @campoy or @mcuadros to move forward with it.

File: examples/language-analyzer.go [hunk 11]
```diff
@@ -6,10 +6,11 @@ import (
 	"io"
 	"os"
 
+	"gopkg.in/src-d/lookout-sdk.v0/pb"
```
[the remark this answers is not in the document]
Comment (smacker) [reply examples/language-analyzer.go]: I asked in a commet above to describe why we need to change the limit above. I see changes but they don't describe the reason. As I mentioned: the only reason is UAST. For a 1mb file it can be huge.
it means we *need* it only in data server but we *can* do it anywhere.

lookoutd uses an internal helper. That helper sets logging, will set metadata and so on. It also changes the limit if I'm not mistaken because of dummy analyzer. Dummy uses the same helper.

File: README.md [hunk 12]
```diff
@@ -1,117 +1,121 @@
...
+- **proto [definitions](./proto)**.
+- pre-generated libraries for [Golang](./pb) and [Python](./python), offering:
+  - an easy **access to the DataService API though a gRPC service**. Lookout will take care of dealing with Git repositories, UAST extraction, programming language detection, etc.
+  - low-level **helpers to work around some protobuf/gRPC caveats**.
+- quickstart [examples](./examples) of an Analyzer that detects language and number of functions (written in Go and in Python).
+
+
+# Caveats
+
+For the gRPC client and server please to follow these requirements:
```
Comment (carlosms) [inline README.md]:
```suggestion
For the gRPC client and server please follow these requirements:
```

File: README.md [hunk 13]
```diff
@@ -1,117 +1,121 @@
...
+- to set a shared maximum gRPC message size in gRPC servers and clients (to avoid hitting different gRPC limits [see [grpc/grpc#7927](https://github.com/grpc/grpc/issues/7927)]); the `Analyzer` service should do so when dialing `DataService` and serving `NotifyReviewEvent` and `NotifyPushEvent`; using lookout-sdk helpers it can be done:
```
Comment (carlosms) [inline README.md]: This renders with too many parenthesis and brakets: "servers and clients (to avoid hitting different gRPC limits [see grpc/grpc#7927])"

What about:

```markdown
For the gRPC client and server please to follow these requirements:
- set a common maximum gRPC message size in gRPC servers and clients. This is required to avoid hitting different gRPC limits, see [grpc/grpc#7927](https://github.com/grpc/grpc/issues/7927). To do so use the included helpers in lookout-sdk:
  - go: using `pb.NewServer` and `pb.DialContext`.
  - python: using `lookout.sdk.grpc.create_channel`.
- support [RFC 3986 URI scheme](https://github.com/grpc/grpc-go/issues/1911); lookout-sdk includes helpers for this:
  - go: using `pb.ToGoGrpcAddress` and `pb.Listen`.
  - python: using `lookout.sdk.grpc.to_grpc_address`.
```

File: README.md [hunk 14]
```diff
@@ -1,117 +1,121 @@
...
+  </a>
+  <a href="https://travis-ci.org/src-d/lookout-sdk">
+    <img src="https://travis-ci.org/src-d/lookout-sdk.svg?branch=master"
+         alt="Build Status">
+  </a>
+  <a href="https://godoc.org/gopkg.in/src-d/lookout-sdk.v0/pb">
+    <img src="https://godoc.org/gopkg.in/src-d/lookout-sdk.v0?status.svg"
+         alt="GoDoc">
+  </a>
+  <a href="https://docs.google.com/document/d/1pqz-_SHO5BsJE-aa8o_bAY3r5vR67amnWN8-qZc2UgY">
```
Comment (carlosms) [inline README.md]: Why the link to the design doc? Did you take this from some template/other project?

I don't really like it. Since it's at the same level as the godoc reference some people may follow it expecting reference material. The design doc was useful for the planning phase, but now we are not keeping it up to date with the latest decisions and features. Also, this particular doc is not even well formatted, I would not link it at the same level as the godoc reference... I would not link it at all.

File: README.md (hunk 3 above)
Comment (carlosms) [reply README.md]: I prefer the html headers, but @mcuadros was quite clear on rejecting its use.
I'd say the safest route would be to keep `lookout-sdk` without html, since it's a lib, and use html only in lookout.

Comment (dpordomingo) [reply README.md]: I understood that **source{d} lookout** would be considered as a whole product, regardless it is separated in different repos, apps, libraries...
But, I'm ok with drop the HTML, that is hard to maintain, and change it again if @src-d/product requires it :dagger:

Comment (dpordomingo) [reply README.md]: I understood that **source{d} lookout** would be considered as a whole product, regardless it is separated in different repos, apps, libraries...
But, I'm ok with drop the HTML, that is hard to maintain, and change it again if @src-d/product requires it :dagger:

**Comment by smacker** [comment]
sorry for the late comment but I just opened readme once again and I don't see much value in toc. The readme is small enough now that toc becomes just a noise for me.
(no need to take any actions in this PR but if you think the same we can update it later)
